source/module_base/math_bspline.cpp

```diff
@@ -1,64 +1,65 @@
 #include "math_bspline.h"
-#include <cstdlib>
+
 #include <assert.h>
+#include <cstdlib>
 
 namespace ModuleBase
 {
-    Bspline::Bspline()
-    {
-        bezier = NULL;
-        norder = 0;
-        xi = 0;
-        Dx = 1.0; 
-    }
-    Bspline::~Bspline()
-    {
-        if(bezier!=NULL) delete[] bezier;
-    }
+Bspline::Bspline()
+{
+    bezier = NULL;
+    norder = 0;
+    xi = 0;
+    Dx = 1.0;
+}
+Bspline::~Bspline()
+{
+    if (bezier != NULL)
+        delete[] bezier;
+}
 
-    void Bspline::init(int norderin, double Dxin, double xiin)
+void Bspline::init(int norderin, double Dxin, double xiin)
```

Comment:
Is is better to move contents of `init` into its constructor? Same for `cleanup` to destructor.

Reply:
We want to reuse the same object of this class. Then it is necesarry. For examle:
Bspline a;
a.init(10,2,3);
do something
a.cleanup();
a.init(8,1,3);
do something
a.cleanup();

Reply:
> We want to reuse the same object of this class. Then it is necesarry. For examle: Bspline a; a.init(10,2,3); do something a.cleanup(); a.init(8,1,3); do something a.cleanup();

I see. 
However, is there any relationship between `a.init(10,2,3);` and `a.init(8,1,3);`, for example the results generated will be used afterwards? If there are no relationship between them(I think it is), I suggest separating them to `Bspline a,b`, for reusing the same object might not be intuitive, and even potentially hazardous.

Further more, the code will be more clear in logic flow while preventing misuse. Afterwards, you only need to do something like `Bspline a(1,2,3); do something;`: no init, no cleanup, no manually allocate, free(!) spaces, all handled with constructor and destructor.

Reply:
The cleanup function has exactly the same contents as destructor, so I think it is fine to delete it. It is actually not called and used in ABACUS.

Reply:
But I think we need to retain the function init. The reason is that here we need to assert that norder is an even and positive integer. I think it is inappropriate to do assertions in a constructor. What's you opinion? @caic99 @Qianruipku